server/src/workspaceLookup.ts

```diff
@@ -0,0 +1,192 @@
+import { WorkspaceFolder, RemoteWorkspace, FileChangeType, TextDocumentChangeEvent, DidChangeWatchedFilesParams } from 'vscode-languageserver';
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import { parseOmtFile, parseOmtText } from './omtFileParser';
+import { globPromise } from './globPromise';
+import { WorkspaceModules } from './workspaceModules';
+
+/**
+ * Tracks changes to the workspace and important files for core functionality of the language server
+ */
+export class WorkspaceLookup {
+    private folders = new Map<string, WorkspaceFolder>();
+    private workspaceModules = new WorkspaceModules();
+    private readonly omtPattern = '**/*.omt';
+
+    /**
+     * The folders being watched for changes.
+     */
+    get watchedFolders() {
+        return Array.from(this.folders.values());
+    }
+
+    /**
+     * All the OMT modules in the watched folders
+     */
+    get watchedModules() {
+        return Array.from(this.workspaceModules.modules.values());
+    }
+
+    /**
+     * Create a new workspaceLookup listening to workspace events.
+     * call `init()` to start listening to these events.
+     * @param workspace workspace of the client
+     */
+    constructor(private workspace: RemoteWorkspace) {
+        this.workspaceModules = new WorkspaceModules();
+    }
+
+    /**
+     * Attach change handlers to workspace folders and open files.
+     * Then do a first scan of all the folders and files for recognized OMT content
+     */
+    init(): Promise<void> {
+        this.workspace.onDidChangeWorkspaceFolders(event => {
+            // This event is fired when the user adds or removes a folder to/from their workspace
+            this.collectionPromise(event.added, this.addFolder)
```

Comment:
I think this would be easier, unless I'm missing something. Also, should these be chained or awaited? Or is it fine if they run in the background?

```suggestion
            Promise.all(event.added.map(this.addFolder));
```

Reply:
This would make the `this.collectionPromise` obsolete

Reply:
so what happened here is that I used `collectionPromise` on multiple places in the code, and then actually checked how it resolved or rejected as a single void result. 

it should also be fine that the scanning of folders runs in the background, it is very unlikely the user has a file open that refers to something in a newly added folder. If they do, the only thing they have to do is switch to another document and back and the processing of the workspace folder will likely be done.

I don't think the signature of `onDidChangeWorkspaceFolders` supports waiting on a promise.

Reply:
so technically we don't need the `Promise.All` here either